### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by jordwalke**
I started with two goals:
- Eliminate log spew when compiling php/hack, and get rid of our old legacy hack raw stubs in the .js files.
- Fix the fast hashing feature. We previously weren't detecting when we needed to rebuild when dependencies changed or when externals/string constants were changed in modules.

In order to implement the first feature, it was easiest to actually just clean up the backend so that implementing backends is mostly just a matter of implementing a consistent signature. Nice side effect is that it makes it easier for people to add new backends, though this was not the original goal.

### CODE REVIEW

File: compiler/lib/backend.re [hunk 1]
```diff
@@ -0,0 +1,175 @@
+open Stdlib;
```
Comment (robz) [inline compiler/lib/backend.re:1]: I was struggling a bit to implement some integration with the python backend. This should make that much easier!

**Review by robz — APPROVED** [review]

**Comment by Lupus** [comment]
> implementing backends is mostly just a matter of implementing a consistent signature

Sounds awesome! Along with several backends which can serve as examples that should get people started.

**Comment by jordwalke** [comment]
@Lupus (or anyone else) Feel free to add a plain-php backend (right now it outputs Hack PHP). It's mostly just a matter of changing some stubs and using `(object)[array]` instead of `Vector` - which is mostly just a matter of changing `php_output.re`, and adding a new backend like demonstrated in this diff (copying the Hack one, and just having it use a different printer).

**Comment by Lupus** [comment]
> @Lupus (or anyone else) Feel free to add a plain-php backend (right now it outputs Hack PHP).

Unfortunately I'm not interested in PHP backend, and don't even know this language :) I'm thinking about Lua backend, Ceph has a way of running Lua code right on the OSD (a process, holding the data), kind of stored procedures, which is pretty neat. But writing in plain Lua is not that encouraging, so far I got some success with [Amulet](https://amulet.works/), which is ML inspired language, which compiles to Lua. The project is pretty experimental, but seems to work for my somewhat Lua-embedded use case. The main drawback is that I can't share types and serialization code between main OCaml program, which talks to Ceph, and code that runs inside Ceph. Having Lua backend for OCaml will enable me to just use atdgen and be sure that types and encoding/decoding are aligned between the two.

Also I'm thinking about Golang backend, we have some codebase in Go, that hits the expressiveness limits of the language, but it works on several platforms, uses a lot of batteries from the rich Go ecosystem, and rewriting that to pure OCaml would be a bad thing to do. But transpiling OCaml to Go looks promising, the cost of writing bindings seems to be fair for the benefits OCaml provides.

My understanding is that building the AST mapper itself is not the biggest part of proper backend, but the runtime library? Is there some minimal subset that needs to be implemented to be able to build and run some hello world application?

Also... Golang is running on multiple threads currently with the ability to detach some closure and let it run concurrently. This can probably be expressed with some magical bindings that look like normal OCaml externals. Attributes do not survive in the bytecode, right?

While I do have interest in this, I don't have time unfortunately to fuel this interest... Hopefully I'll find it one day :)

**Pull request merged**